[ELY-2112] Automatic registration of client side / JVM wide default SSLContext

skyllarr opened the pull request:
https://issues.redhat.com/browse/ELY-2112

Adds Elytron client provider that provides default SSLContext which can be returned by `SSLContext.getDefault()` call. The default SSLContext is considered to be the one that matches all rules.

skyllarr committed:
[ELY-2112] Automatic registration of client side / JVM wide default SSLContext

darranl commented on auth/client/src/main/java/org/wildfly/security/auth/client/DefaultSSLContextSpi.java:
```diff
@@ -0,0 +1,72 @@
+package org.wildfly.security.auth.client;
```
Missing copyright header.

Skyllarr replied:
Fixed. Thank you

darranl commented on auth/client/src/main/java/org/wildfly/security/auth/client/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,67 @@
+package org.wildfly.security.auth.client;
```
Another missing header.

Skyllarr replied:
Missing copyrights added.

darranl commented on auth/client/src/main/java/org/wildfly/security/auth/client/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,67 @@
+        public Object newInstance(Object ignored) throws NoSuchAlgorithmException {
+            Integer enteredCountTmp = entered.get();
+            entered.set(enteredCountTmp == null ? 1 : enteredCountTmp + 1);
+            if (entered.get() >= 2) {
```
I would suggest include this info in a short comment in the file.

Skyllarr replied:
Comment added

darranl commented on auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContext.java:
```diff
@@ -244,8 +244,15 @@ public AuthenticationContext withoutSsl(int idx) throws IndexOutOfBoundsExceptio
-            if (node.getRule().matches(uri, abstractType, abstractTypeAuthority)) return node;
-            node = node.getNext();
+            if (uri == null) {
+                if (node.getRule().equals(MatchRule.ALL)) {
```
I assume this is needed here as the default SSLContext implementation doesn't know the destination it is being used for so we have no URI - are there any compatibility issues by matching ALL possibly sooner than we used to or do we feel this is safe?

Skyllarr replied:
Matching ALL sooner than we used to would happen only if null URI was passed here. Which was not possible before my changes because this method is only called in [getSSLContextFactory](https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContextConfigurationClient.java#L235) and there was a [null check for URI](https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContextConfigurationClient.java#L233). I removed this null check in this PR. 

I think backwards incompatibility can happen only if users were relying on the IllegalArgumentException to happen by passing null to the [getSSLContextFactory](https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContextConfigurationClient.java#L232) method. Is this something we should be worried about?

Btw to avoid this I can introduce new config parameter for default SSL context meant only for this provider. But that would not leverage existing possibility to configure default SSLContext.

darranl commented on auth/client/src/main/java/org/wildfly/security/auth/client/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,67 @@
+                }
+                // if we had an exception previously, then default ssl context was still returned from other security provider
+                // which is why we need to check entered variable again
+                if (entered.get() >= 2) {
```
Also here, we may want to include this info in a comment.

Skyllarr replied:
Comment added.

darranl reviewed (dismissed):
No major issues, just some missing copyright headers and suggestions about comments.

fabiobrz commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextFromFileWorksAndHasPrecedenceTestCase.java:
```diff
@@ -0,0 +1,29 @@
+        Security.insertProviderAt(new ElytronClientDefaultSSLContextProvider(CONFIG_FILE), 1);
+        Assert.assertNotNull(Security.getProvider("ElytronClientDefaultSSLContextProvider"));
+        AuthenticationContext.empty().run(() -> {   // provider will not use current auth context but will use authentication context from the file passed to provider. File passed to provider has precedence
+            SSLContext cip = null;
```
Just for personal knowledge (but maybe worth a renaming), what does `cip` stand for?

Skyllarr replied:
@fabiobrz sorry it was a bad name, originally it meant `context in provider` but that was just bad naming. I changed it to `default SSLContext`.

fabiobrz replied:
Thanks

fabiobrz commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextFromFileWorksAndHasPrecedenceTestCase.java:
```diff
@@ -0,0 +1,29 @@
+                Assert.fail();
+            }
+            Assert.assertNotNull(cip);
+            Assert.assertEquals(cip.getProvider().getName(), ElytronClientDefaultSSLContextProvider.class.getSimpleName());
```
IIUC, in case this is the "has precedence" part - and the following one is the "works" part - then I'd switch the two lines' order so that it matches the test method name.

Skyllarr replied:
@fabiobrz switched

fabiobrz commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextProviderCanDelegateWhenConfigIsMissingTestCase.java:
```diff
@@ -0,0 +1,28 @@
+    @Test
+    public void testDefaultSSLContextProviderWillDelegateIfNoConfigured() {
+        Security.insertProviderAt(new ElytronClientDefaultSSLContextProvider(), 1);
+        Assert.assertEquals(ElytronClientDefaultSSLContextProvider.class.getSimpleName(), Security.getProvider("ElytronClientDefaultSSLContextProvider").getName());
```
IIUC, here you mean to check that the inserted provider is the `ElytronClientDefaultSSLContextProvider`, which seems right to me but in the previous test class you just checked for the named one not to be null. Can we have a single way to assess that the provider registration was successful and consistent?

Skyllarr replied:
@fabiobrz If the provider is not null it is certain that this equality is true, so I left the assertNotNull check.  And I changed this line to be consistent everywhere.

fabiobrz replied:
Great, thanks!

fabiobrz commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextProviderDoesNotLoopTestCase.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Test
+    public void testDefaultSSLContextProviderDoesNotLoopTestCase() throws GeneralSecurityException, URISyntaxException, ConfigXMLParseException {
+        Security.insertProviderAt(new ElytronClientDefaultSSLContextProvider(), 1);
```
Isn't there the need to check that the registration went well as it is done in the previous test cases?

Same question about `DefaultSSLContextProviderFromCurrentAuthContextTestCase`.

Skyllarr replied:
@fabiobrz Added, thanks.

fabiobrz commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextProviderFromCurrentAuthContextTestCase.java:
```diff
@@ -0,0 +1,33 @@
+    @Test
+    public void testDefaultSSLContextTakenFromCurrentContext() throws GeneralSecurityException, URISyntaxException, ConfigXMLParseException {
+        Security.insertProviderAt(new ElytronClientDefaultSSLContextProvider(), 1);
+        AuthenticationContext authenticationContext = ElytronXmlParser.parseAuthenticationClientConfiguration(new URI(CONFIG_FILE)).create();
```
I'd add a Javadoc comment to this test method so that it is clear what it does, as you did in the previous ones.

BTW could you please convert the above mentioned (existing) comments to Javadoc as well?

Skyllarr replied:
@fabiobrz I added javadoc on top of class in each test. I left in-line comments when I though it migh be useful to specify exact line. Thanks

fabiobrz reviewed (changes requested):
Hi @Skyllarr - left my review with some comments. 

On a general note, I couldn't find tests for the following cases (from the feature test plan security testing section). Am I missing something?

* Test that exception will be thrown when trying to set invalid configuration path to the provider
* Test that exception will NOT be thrown when not providing configuration path and wildfly-config.xml is not found on the classpath.

Skyllarr commented:
> Hi @Skyllarr - left my review with some comments.
> 
> On a general note, I couldn't find tests for the following cases (from the feature test plan security testing section). Am I missing something?
> 
> * Test that exception will be thrown when trying to set invalid configuration path to the provider
> * Test that exception will NOT be thrown when not providing configuration path and wildfly-config.xml is not found on the classpath.

I added missing test. For ` Test that exception will be thrown when trying to set invalid configuration path to the provider` please see DefaultSSLContextNonexistentConfigFileTest.java
For `Test that exception will NOT be thrown when not providing configuration path and wildfly-config.xml is not found on the classpath.` please see DefaultSSLContextProviderWillDelegateWhenConfigIsMissingTest.java

Skyllarr commented on auth/client/src/main/java/org/wildfly/security/auth/client/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public ElytronClientDefaultSSLContextProvider(String configPath) {
+        super("ElytronClientDefaultSSLContextProvider", 1.0, "Elytron client provider for default SSLContext");
```
@fabiobrz The `super(String, String, String)` constructor was introduced in JDK 9, so I canot use it here because we have release configured for compiler to be 8. The deprecated constructor I used is still present in JDK 15 so I used it but good point, I will check today with the team whether the release can be changed to 9. If yes I can change it, if not I have to leave it as it is. I will let you know

fabiobrz replied:
Thank you very much @Skyllarr - not a big issue, though. There is valid reasoning already for leaving it as is, which you provided.

I'll leave the final call to you here.

Skyllarr replied:
@fabiobrz Yes we have to use the deprecated constructor here, as wildfly should work with JDK 8.

Skyllarr commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextFromFileWorksAndHasPrecedenceTestCase.java:
```diff
@@ -0,0 +1,29 @@
+    private static final String CONFIG_FILE = "file:./src/test/resources/org/wildfly/security/auth/client/test-wildfly-config-client-default-sslcontext.xml";
+
+    @Test
+    public void testDefaultSSLContextFromFileWorksAndHasPrecedence() {
```
@fabiobrz Added assertion fail messages

fabiobrz replied:
At least to most relevant ones :slightly_smiling_face: ...

But yeah, that makes sense, the remaining three assertions are clear enough and don't change among the tests, so I am okay with this.

Skyllarr commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextFromFileWorksAndHasPrecedenceTestCase.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Test
+    public void testDefaultSSLContextFromFileWorksAndHasPrecedence() {
+        Security.insertProviderAt(new ElytronClientDefaultSSLContextProvider(CONFIG_FILE), 1);
```
@fabiobrz Provider takes string as an argument because this argument can be passed to the provider in java.security file like this:

```
java.security.provider.1=org.wildfly.security.auth.client.ElytronClientDefaultSSLContextProvider /path/to/config/file
```

In above cases the JVM is parsing the java.security file and the argument passed to the provider is considered to be String.

I can add another constructor with URI, but it would be usable only from code and I think this provider would usually be configured in the java.security. WDYT? I will check this with the team today as well.

fabiobrz replied:
> In above cases the JVM is parsing the java.security file and the argument passed to the provider is considered to be String.

Ah thanks, that sounds reasonable to me to me. 

> I can add another constructor with URI, but it would be usable only from code and I think this provider would usually be configured in the java.security. WDYT? I will check this with the team today as well

Eh... The only thing that I didn't actually feel to be correct is that usage of hard coded paths in test classes which would warrant for the overloaded ctor actually, but as said you'd have to overload it in a couple classes more, if I don't go wrong.

Skyllarr replied:
@fabiobrz If I add the constuctor with  URI (yes it would require overloading other classes as well), in tests I would still want to use the String ctor also, as that one would need testing the most anyway.

fabiobrz replied:
Yep, thanks, let's move on like this since it is a test use case.

fabiobrz commented:
> @fabiobrz I added missing test. For ` Test that exception will be thrown when trying to set invalid configuration path to the provider` please see DefaultSSLContextNonexistentConfigFileTest.java
> For `Test that exception will NOT be thrown when not providing configuration path and wildfly-config.xml is not found on the classpath.` please see DefaultSSLContextProviderIsIgnoredWhenConfigIsMissingTest.java

ok, thanks - I somehow missed the expected exception declaration, my bad, sorry.

Skyllarr commented:
@fabiobrz Thanks for your review, I replied to the remaining comments. Let me know what you think.

fabiobrz reviewed (approved):
Hi @Skyllarr, and thanks for addressing my comments and questions.
The tests development part LGTM, so I am approving the PR with respect to that.

fjuma commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextFromFileWorksAndHasPrecedenceTest.java:
```diff
@@ -0,0 +1,34 @@
+package org.wildfly.security.auth.client;
```
Missing header.

fjuma commented on auth/client/src/test/java/org/wildfly/security/auth/client/DefaultSSLContextNonexistentConfigFileTest.java:
```diff
@@ -0,0 +1,34 @@
+package org.wildfly.security.auth.client;
```
Same here.

Skyllarr replied:
@fjuma Thank you for review. I added missing headers. I also switched values in AssertEqual calls because expected and actual values were switched.

Skyllarr commented:
I have put this on hold now. During manual testing I found that the provider registration from java.security file does not work for me, not sure if the fact that this provider is unsigned is the reason.

Skyllarr commented:
> I have put this on hold now. During manual testing I found that the provider registration via java.security file does not work for me, not sure if the fact that this provider is unsigned is the reason.

Just a quick update, this has nothing to do with signing, as the provider does not provide encryption it does not need to be signed. Reason it fails when registering it through `java.security` file is that if I add just `wildfly-elytron-client` jar to the JVM's external libraries the provider will throw NoClassDefFound error for class `ConfigXMLParseException`.  But if I add `wildfly-client-all` jar to external libraries of JVM, the JVM is unable to start.

Btw I tried to remove scope provided for `wildfly-client-config`  to avoid NoClassDefFound, but it did not help the Exception.

skyllarr commented on auth/client-provider/pom.xml:
```diff
@@ -0,0 +1,91 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
```
Let me know if the  location of this new module and the name of the folder `client-provider` (with `-` sign) is OK.

skyllarr commented on auth/client-provider/src/test/resources/org/wildfly/security/auth/client/provider/test-wildfly-config-client-default-sslcontext.xml:
```diff
@@ -0,0 +1,30 @@
+    <authentication-client xmlns="urn:elytron:client:1.5">
+        <key-stores>
+            <key-store name="keystore1" type="JKS">
+                <file name="src/test/resources/client.keystore"/>
```
Let me know if I should generate this keystore instead. It is not actually used to connect in the test I just needed some path here. The client.keystore is copied from [test folder of auth/client.](https://github.com/wildfly-security/wildfly-elytron/blob/1.x/auth/client/src/test/resources/client.keystore)

fjuma replied:
Just to keep things simpler, if it is possible to generate it instead that would be a cleaner approach.

skyllarr commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/DefaultSSLContextSpi.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Source: A utility method taken from https://github.com/wildfly/wildfly-client-config/blob/master/src/main/java/org/wildfly/client/config/ClientConfiguration.java on March 2022
+     */
+    static URI configPathUrlToUri(String wildFlyConfig) {
```
I added this method in the most recent changes. It is taken from [wildfly-client-config repository.](https://github.com/wildfly/wildfly-client-config/blob/master/src/main/java/org/wildfly/client/config/ClientConfiguration.java) Previously I had to specify schema `file:` at the beginning of wildfly-config.xml path passed to the provider in both the code and in the `java.security` file. Thanks to this method `file:` does not have to be at the beginning and it also deals with windows paths.

skyllarr commented on auth/client-provider/assembly.xml:
```diff
@@ -0,0 +1,23 @@
+    </formats>
+    <includeBaseDirectory>false</includeBaseDirectory>
+    <dependencySets>
+        <dependencySet>
```
I added this assembly.xml file used by the `maven-assembly-plugin`  in order to package dependencies with scope provided and create fat jar. The existing `jar-with-dependencies` functionality of `maven-assembly-plugin` would package only the runtime scope.

fjuma commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/DefaultSSLContextSpi.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    public DefaultSSLContextSpi(String configPath) throws GeneralSecurityException, ConfigXMLParseException {
+        this(ElytronXmlParser.parseAuthenticationClientConfiguration(configPathUrlToUri(configPath)).create());
```
I think we need some validation for `configPath` here. Looking in the wildfly-config-client repo, it looks like they check that `configPath` is not `null` and it's not empty before calling the corresponding `propertyUrlToUri` method.

skyllarr replied:
@fjuma Added IllegalArgumentException if the configPath is null or empty. Thanks!

fjuma commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/DefaultSSLContextSpi.java:
```diff
@@ -0,0 +1,122 @@
+
+    private SSLContext configuredDefaultClientSSLContext;
+
+    public DefaultSSLContextSpi() throws GeneralSecurityException {
```
Would be good to add javadoc for the various constructors.

skyllarr commented:
Summary of the changes I did since last time this PR got approved:
 * I added new module `client-provider` to the `auth` subdirectory. It is being placed on the same level as client module. All classes together with test classes and test resources have been placed to this new module
 * I added assembly.xml file which is being used by maven-assembly-plugin in order to package runtime dependencies and dependencies with scope provided and create fat jar. I will double check other options of this plugin if I missed something.
 * I copied a utility method from [wildfly-client-config repository.](https://github.com/wildfly/wildfly-client-config/blob/master/src/main/java/org/wildfly/client/config/ClientConfiguration.java) that takes string path and returns URI. 
 
After these changes when using the provider the jar `wildfly-elytron-client-provider-XYversion.jar` should be placed on the classpath. I will change this in the community docs and proposal.

fjuma commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/DefaultSSLContextSpi.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    public DefaultSSLContextSpi(AuthenticationContext authenticationContext) throws GeneralSecurityException {
+        AuthenticationContextConfigurationClient AUTH_CONTEXT_CLIENT = AccessController.doPrivileged((PrivilegedAction<AuthenticationContextConfigurationClient>) AuthenticationContextConfigurationClient::new);
```
I think we need to make sure `authenticationContext` isn't `null`.

skyllarr replied:
Added

fjuma commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,109 @@
+
+    public static final String ELYTRON_CLIENT_DEFAULT_SSL_CONTEXT_PROVIDER_NAME = "ElytronClientDefaultSSLContextProvider";
+
+    public ElytronClientDefaultSSLContextProvider() {
```
Would be good to add javadoc for the constructors.

skyllarr replied:
Added

fjuma commented on auth/client-provider/src/main/java/org/wildfly/security/auth/client/provider/ElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,109 @@
+ */
+@MetaInfServices(value = Provider.class)
+public final class ElytronClientDefaultSSLContextProvider extends Provider {
+
```
Missing `serialVersionUID`.

skyllarr replied:
Added

fjuma commented on auth/client-provider/src/test/java/org/wildfly/security/auth/client/provider/DefaultSSLContextNonexistentConfigFileTest.java:
```diff
@@ -0,0 +1,53 @@
+import java.security.Security;
+
+/**
+ * Test that default SSLContext provider will throw an exception when configured file path does not exist
```
Should we also add a test for the case where the default constructor (`new ElytronClientDefaultSSLContextProvider()`) is used, i.e., when `CONFIG_FILE` is `null`?

fjuma replied:
Looks like this is actually already covered by `DefaultSSLContextProviderIsIgnoredWhenConfigIsMissingTest`.

fjuma commented on auth/client-provider/src/test/java/org/wildfly/security/auth/client/provider/DefaultSSLContextNonexistentConfigFileTest.java:
```diff
@@ -0,0 +1,53 @@
+        authenticationContext.run(() -> {
+            SSLContext defaultSSLContext = null;
+            try {
+                defaultSSLContext = SSLContext.getDefault();
```
I think this line throws the `IllegalArgumentException`, right?

fjuma commented on auth/client-provider/src/test/java/org/wildfly/security/auth/client/provider/DefaultSSLContextNonexistentConfigFileTest.java:
```diff
@@ -0,0 +1,53 @@
+            }
+            Assert.assertNotNull(defaultSSLContext);
+            Assert.assertNotEquals(ElytronClientDefaultSSLContextProvider.class.getSimpleName(), defaultSSLContext.getProvider().getName()); // diff provider was used since elytron provider is looping
+            Assert.assertNotNull(defaultSSLContext.getSocketFactory());
```
So then these assertions won't actually be hit in this test.

skyllarr replied:
Correct, I removed those assertions

fjuma commented on auth/client-provider/src/test/java/org/wildfly/security/auth/client/provider/DefaultSSLContextProviderProgrammaticConfigurationTest.java:
```diff
@@ -0,0 +1,57 @@
+ * Test that default SSLContext from provider can use programmatic configuration
+ */
+public class DefaultSSLContextProviderProgrammaticConfigurationTest {
+    private static final String CONFIG_FILE = "file:./src/test/resources/org/wildfly/security/auth/client/provider/test-wildfly-config-client-default-sslcontext.xml";
```
Just curious, is the `file:` needed here? Or are we just trying the case where it is specified?

skyllarr replied:
@fjuma Yes exactly, `file:` is not needed. I just left it this way to test this variation as well

fjuma commented on auth/client-provider/src/test/resources/org/wildfly/security/auth/client/provider/test-wildfly-config-client-default-sslcontext.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<configuration>
+    <authentication-client xmlns="urn:elytron:client:1.5">
```
I think the latest schema version is 1.7 so we should update this here and in the other files too.

skyllarr replied:
Fixed

fjuma commented on auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContextConfigurationClient.java:
```diff
@@ -230,7 +234,6 @@ public SSLContext getSSLContext(URI uri, AuthenticationContext authenticationCon
      * @return the matching SSL context factory (not {@code null})
      */
     public SecurityFactory<SSLContext> getSSLContextFactory(URI uri, AuthenticationContext authenticationContext, String abstractType, String abstractTypeAuthority) {
-        Assert.checkNotNullParam("uri", uri);
```
The javadoc for the parameter might need to be updated if it mentions that `null` is not allowed.

skyllarr replied:
Fixed

fjuma commented on auth/client/src/main/java/org/wildfly/security/auth/client/_private/ElytronMessages.java:
```diff
@@ -201,4 +201,10 @@ ConfigXMLParseException xmlUnableToIdentifyProvider(@Param Location location, St
+    @Message(id = 14005, value = "Default SSL context in security provider creates infinite loop")
+    NoSuchAlgorithmException sslContextForSecurityProviderCreatesInfiniteLoop();
+
+    @Message(id = 14007, value = "Configuration file path passed to ElytronClientDefaultSSLContextProvider not found")
```
Should this be 14006? It doesn't look like it's already used.

skyllarr replied:
Fixed

fjuma commented on auth/client-provider/src/test/java/org/wildfly/security/auth/client/provider/DefaultSSLContextFromFileWorksAndHasPrecedenceTest.java:
```diff
@@ -0,0 +1,53 @@
+                Assert.fail("Obtaining of default SSLContext with both config file and programmatic configuration present threw NoSuchAlgorithmException exception.");
+            }
+            Assert.assertNotNull(defaultSSLContext);
+            Assert.assertNotNull(defaultSSLContext.getSocketFactory());
```
It doesn't look like we have any tests that validate the returned default `SSLContext` (e.g., Say we configured a `default-ssl-context` in `wildfly-config.xml` with specific cipher suites, protocols, etc. It would be good to have a test that validates the returned `SSLContext` to make sure it's correct).

skyllarr replied:
@fjuma Added protocol and cipher specification and test checks that. Thanks!

fjuma commented on auth/client-provider/assembly.xml:
```diff
@@ -0,0 +1,23 @@
+<assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.3"
```
Missing `<?xml version="1.0" encoding="UTF-8"?>`

fjuma commented on auth/client-provider/pom.xml:
```diff
@@ -0,0 +1,91 @@
+                </configuration>
+            </plugin>
+            <plugin>
+                <artifactId>maven-assembly-plugin</artifactId>
```
Just curious, is there a reason to use this instead of `maven-shade-plugin`?

fjuma commented:
@Skyllarr Was taking a look at the wildfly-elytron-client-provider JAR and it contains a lot of stuff in it.

Currently, this JAR is only needed for the case where a user is registering the provider in the `java.security` file and attempting to add the JAR to the classpath manually.

What I'm wondering is if instead of producing this JAR with all these dependencies, could we just address this in the documentation instead? For example, we could mention that when registering this provider in the `java.security` file, the user needs to add a runtime dependency on `wildfly-elytron-client` to their client application.

fjuma commented on auth/client/src/main/java/org/wildfly/security/auth/client/AuthenticationContextConfigurationClient.java:
```diff
@@ -196,6 +196,10 @@ private static AuthenticationConfiguration initializeConfiguration(final URI uri
         return configuration;
     }
 
+    public SSLContext getSSLContext(AuthenticationContext authenticationContext) throws GeneralSecurityException {
```
Would be good to add javadoc here.

skyllarr replied:
Added. Thanks!

fjuma commented:
Thanks for the updates @Skyllarr! I just did another pass and have only added a small comment now.

@darranl You had previously reviewed this a little while ago. There have been a bunch of updates since then. If you could take another pass through the updated changes, that would be great. Thanks.

fjuma commented on auth/client/src/main/java/org/wildfly/security/auth/client/WildFlyElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,117 @@
+                if (e.getCause() instanceof FileNotFoundException) {
+                    throw log.clientConfigurationFileNotFound();
+                }
+                throw new NoSuchAlgorithmException(e);
```
Would be good to add to ElytronMessages.

fjuma reviewed (approved):
Thanks @Skyllarr!

skyllarr commented on auth/client/src/main/java/org/wildfly/security/auth/client/WildFlyElytronClientDefaultSSLContextProvider.java:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @param configPath path to Elytron client configuration path
+     */
+    public Provider configure(String configPath) {
```
My latest change was overriding this method so that JDK9+ can handle passing of the config path URL from `java.security` file.

fjuma replied:
Thanks @Skyllarr! Looks good to me.
